### DISCUSSION TITLE
AppImage with wxWidgets breaks because of compositing

### DISCUSSION THREAD

**Discussion by onli** [AppImage/AppImageKit#1300] [Q&A]
I have a dock that uses wxWidgets to build the UI. Recently I enabled compositing, which worked fine on my system. But it doesn't work anymore when starting the dock via the AppImage. My guess is that the old wxWidgets version (3.0 vs 3.2) is to blame, compositing was always broken for me before when trying to enable it in this app, and according to discussion posts about wxWidgets it was known to be buggy. How should I continue now?

 1. Is it a good idea to raise the wxWidgets version used to build the app? 
  1.  If yes: https://packages.ubuntu.com/lunar/libwxgtk3.2-dev seems to be the earliest occurrence of wxWidets-3.2 in Ubuntu. Should I move the base of the appImage to an Ubuntu version so high? Will that even work?
  2.  Or, also if yes: Should I import a newer wxWidgets version into an older Ubuntu version, maybe even into the oldest supported LTS? Has someone done so already?
 2. Alternatively, maybe someone here ran into the same scenario with wxWidgets when creating an AppImage and knows how to workaround this specific bug? It would be acceptable to disable compositing completely in the AppImage (simdock does have working pseudo-transparency as fallback) - but how do you feel about me providing a worsened AppImage? - and I assume it's possible that compositing can somehow be enabled even on wxWidgets 3.0.

My project writeup of this issue is at https://github.com/onli/simdock/issues/16, I'd be happy for feedback either here or there :)

**Comment by probonopd** [comment]
Thanks @onli for making an AppImage of simdock and thanks for asking.

> simdock when run from the AppImage stays black when compositing is on

So what you are encountering seems to be not so much an issue with the AppImage "self-mounting disk image" format itself but with the deployment tools used to produce the AppDir that goes inside the AppImage.

> So I switched to https://github.com/linuxdeploy/linuxdeploy

linuxdeploy uses plugins to do the deployment, I think there is one for Gtk but I am not so sure about wxWidgets. Possibly a custom plugin would be needed for it, but I cannot say for sure without knowing what actually causes the error.

> At first I though that AppImage might block access 

AppImage doesn't block access to anything, but some applications try to load resources from hardcoded paths within `/usr` - this won't work with AppImages because the files are inside the AppImage and not within `/usr`. But since you say that it works when compositing is off, this doesn't seem to be causing the issue you are running into.

> AppImage breaks compositing (unlikely, given how AppImage works)

AppImage: unlikely the cause. Making the AppDir that goes into the AppImage: could well be the culprit. One way to try out: Run the AppRun file in your AppDir directly, without using the AppImage. Possbily something breaks if some paths can't be resolved properly. Do you know how to use `strace`? Perhaps you can find out something withen you run `strace ./AppDir/AppRun | grep ENOENT` or similar.

> The old wxWidget version used for the AppImage is broken.

Well. That's easy to rule out: Just build and install simdock "normally" on your build system, and see whether it works when compiled and installed on that system "traditionally". (E.g., using a Ubuntu Live ISO)

Next step: See if we can raise the wxWidgets version - maybe by using a newer Ubuntu version as base? 

> Conceptually the right thing to do would be to use an Ubuntu version no newer than the currently oldest still-supported LTS (20.04 at the time of this writing), and build a newer wxWidgets version on that. I'd only spend this effort before having tested the previous point, though.

Hope this helps!

**Reply by TheAssassin** [reply]
Indeed, there is a chance a wxWidgets plugin for linuxdeploy could help here, but I'm not entirely sure whether they have any additional resources that need to be bundled anyway (stuff your own application loads usually needs to be handled by yourself anyway).

Contributions welcome!

**Reply by onli** [reply]
> but I'm not entirely sure whether they have any additional resources that need to be bundled anyway

My guess is that since the app can compile and start, that it's unlikely that a dependency is missing. The GTK plugin seems to have worked for that, primarily by enabling Wnck (which my program also uses). 

Though I was surprised that linuxdeploy did not detect everything automatically that is needed to make the program run, when creating the AppImage with it. I guess that is what the appimage-builder would do instead? And linuxdeployqt worked for that more by accident in the past?

> Well. That's easy to rule out: Just build and install simdock "normally" on your build system, and see whether it works when compiled and installed on that system "traditionally". (E.g., using a Ubuntu Live ISO).

It's good to get ideas like that offered :) Seems like the way forward. I'll test that next, thanks!

**Reply by onli** [reply]
Indeed: The test shows the same problem on an older Ubuntu. Compositing leads to a grey background (light grey instead of black grey). I used 22.04 as that was installed on a system in the house, but 22.04 is also still on wxWidgets 3.0.5. 20.04 uses 3.0.4, while my main system where it worked uses 3.2.4. I think that's close enough to confirm the problem as related to wxWidgets, not AppImage directly.

Time to try updating wxWidgets on 20.04 before compiling - or I could try to update wxWidgets via the package from 23.04 on 22.04 LTS -, or should I try to build the AppImage with Ubuntu 23.04 instead (that one has 3.2.2 in its repo), maybe with the appimage-builder?

**Reply by probonopd** [reply]
Well, just for testing you could make the AppImage on the most recent Ubuntu. That AppImage won't run on all currently-still supported releases of Ubuntu but it may nevertheless be useful for your testing/debugging.

If it works there, then the next step would be to find/build a newer wxWidgets for Ubuntu 20.04 and make an AppImage using that. That AppImage should then run on all currently-still supported releases of Ubuntu.
